Bump WordPress and PHP minimum

peterwilsoncc opened the pull request:
<!--
Filling out this template is required.  Any PR that does not include enough information to be reviewed may be closed at a maintainers' discretion.  All new code requires documentation and tests to ensure against regressions.
-->

### Description of the Change
<!--
We must be able to understand the design of your change from this description.  The maintainer reviewing this PR may not have worked with this code recently, so please provide as much detail as possible.

Where possible, please also include:
- verification steps to ensure your change has the desired effects and has not introduced any regressions
- any benefits that will be realized
- any alternative implementations or possible drawbacks that you considered
- screenshots or screencasts
-->

Bump WordPress minimum to 5.7
Bump PHP minimum to 7.4

<!-- Enter any applicable Issue number(s) here that will be closed/resolved by this PR. -->
Closes #905, Closes #925

### How to test the Change
<!-- Please provide steps on how to test or validate that the change in this PR works as described. -->



### Changelog Entry
<!--
Please include a summary for this PR, noting whether this is something being Added / Changed / Deprecated / Removed / Fixed / or Security related.  You can replace the sample entries after this comment block with the single changelog entry line for this PR. -->
> Changed - Now requires PHP 7.4 or later and WordPress 5.7 or later.


### Credits
<!-- Please list any and all contributors on this PR so that they can be added to this projects CREDITS.md file. -->
Props @peterwilsoncc, @vikrampm1, @iamdharmesh.


### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you are unsure about any of these, please ask for clarification.  We are here to help! -->
- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests pass.

peterwilsoncc committed:
Update readme files with minimum supported versions.

peterwilsoncc committed:
Update workflows using PHP or WP versions.

peterwilsoncc committed:
Remove redundent applications password plugin.

peterwilsoncc committed:
Prevent activation on old versions.

peterwilsoncc committed:
Update PHP versions in docunit file.

peterwilsoncc commented on distributor.php:
```diff
@@ -55,14 +55,29 @@ function( $class ) {
- * Require PHP version 5.6 - throw an error if the plugin is activated on an older version.
+ * Require PHP 7.4+, WP 5.7+ - throw an error if the plugin is activated on an older version.
  */
 register_activation_hook(
```
These changes prevent activation of the plugin on unsupported versions of WP, PHP.

There is now the small risk a site will have an active plugin and update to a version it no longer supports. How have we handled that in the past?

dkotter replied:
So I was trying to find a version of this in our plugins and couldn't quite remember where it's at but what we've done in the past is have some code in the main plugin file (so this file in this case) that checks if our requirements are met and if not, output an admin notice and stop execution of the plugin. Not our plugin but here's a similar approach in a Woo plugin: https://github.com/woocommerce/google-listings-and-ads/blob/develop/google-listings-and-ads.php#L41 and https://github.com/woocommerce/google-listings-and-ads/blob/develop/src/Internal/Requirements/VersionValidator.php#L22

peterwilsoncc replied:
In 6187c5625d1f612a7715b3be0c9da9bf0c725c7a I've added a first pass at showing a notice if the plugin fails to meet minimum version requirements.

Honestly, it seems like a lot for something that ought to be quite minor but I am not sure if there is much that can be done about that. For this to work, the plugin's main file will always need to be compatible with earlier versions of WP/PHP. The up-to-date code will need to come later.

Should I move some of the bootstrapping code so `distributor.php` can focus on activation & a bootstrapper can focus on bootstrapping?

peterwilsoncc committed:
Use two letter version numbers for version compare.

peterwilsoncc committed:
`wp_is_application_passwords_available()` exists in all supported versions.

peterwilsoncc committed:
Simplify `is_using_gutenberg()` for WP5.7+ support only.

peterwilsoncc committed:
Merge branch 'develop' into fix/requirement-bump

# Conflicts:
#	readme.txt

peterwilsoncc committed:
Add mocked functions required for updated Gutenberg detection.

peterwilsoncc committed:
Show notice if version requirements fail.

peterwilsoncc committed:
Remove note about application password merge.

peterwilsoncc committed:
Refer to application passwords feature rather than plugin.

peterwilsoncc committed:
Improve strings for minimum requirements.

peterwilsoncc committed:
Move bootstrapping to dedicated file.

peterwilsoncc committed:
Use internal helper function to determine if post type is supported.

Expand explanation of internal helper to document why it exists.

peterwilsoncc committed:
Delay calling `is_using_gutenberg` until `use_block_editor_for_post_type` is defined.

peterwilsoncc commented on includes/syndicated-post-ui.php:
```diff
@@ -26,12 +28,8 @@ function() {
-
-			if ( $post && ! \Distributor\Utils\is_using_gutenberg( $post ) ) {
-				add_action( 'do_meta_boxes', __NAMESPACE__ . '\replace_revisions_meta_box', 10, 3 );
-				add_action( 'add_meta_boxes', __NAMESPACE__ . '\add_revisions_meta_box' );
-			}
+			add_action( 'do_meta_boxes', __NAMESPACE__ . '\replace_revisions_meta_box', 10, 3 );
+			add_action( 'add_meta_boxes', __NAMESPACE__ . '\add_revisions_meta_box' );
```
The `is_using_gutenberg()` function call is moved to the `do_meta_boxes` hook where it is used. This ensures all the editor functions are defined before attempting to use them. Specifically, in this instance, `use_block_editor_for_post_type()`.

peterwilsoncc committed:
Add sniff for main file’s bespoke compatibility requirements.

peterwilsoncc committed:
Document main plugin file’s bespoke compatibility requirements.

peterwilsoncc committed:
Remove application password plugin reference in admin.

peterwilsoncc committed:
Include authentication endpoint in distributor endpoint.

peterwilsoncc committed:
Defer to `use_block_editor_for_post_type()` if it exists.

Instead of reproducing the function (including the final filter) call the function if it exists. Otherwise go through the polyfill steps.

peterwilsoncc committed:
Merge branch 'develop' into fix/requirement-bump

# Conflicts:
#	distributor.php

peterwilsoncc committed:
Avoid deprecated `gutenberg_get_jed_locale_data()`.

iamdharmesh reviewed (commented):
Thanks a lot for the great work here @peterwilsoncc, Code looks good to me just want you to check on the notes below.

- I am getting 404 for admin.min.css
![image](https://user-images.githubusercontent.com/10613171/185553248-91ac7adc-b895-4450-8ba1-c81f1929f51a.png)

- I think we should also add minimum requirements in the plugin header of the plugin’s main PHP file as well. Since WordPress 5.8 plugin [readme files are not parsed for requirements](https://core.trac.wordpress.org/ticket/48520). This means that headers `Requires PHP` and `Requires at least` are going to be parsed from plugin’s main PHP file.

Thanks

peterwilsoncc committed:
Add minimum requirements to main plugin file.

peterwilsoncc committed:
Update package-lock file for npm 8.x.

peterwilsoncc commented:
@iamdharmesh I've added the requirements to the main plugin file in c1c93fb21b01a57cff5c7d39dc7ac10f360dc732 

You'll need to run `npm i; npm run build` in your command line for the css files to be generated.

iamdharmesh commented:
Thanks for adding requirements to the main plugin file @peterwilsoncc.

> You'll need to run `npm i; npm run build` in your command line for the css files to be generated.

Yes, I am already doing this, but it seems the URL to load CSS file is wrong. it is `.../distributor/includes/dist/css/...` instead of `.../distributor/dist/css/...`.

Thanks

iamdharmesh commented on includes/bootstrap.php:
```diff
@@ -0,0 +1,278 @@
+		}
+
+		// Dependencies only apply to JavaScript, not CSS files.
+		wp_enqueue_style( 'dt-admin-settings', plugins_url( '/dist/css/admin.min.css', __FILE__ ), array(), $asset_data['version'] );
```
```suggestion
		wp_enqueue_style( 'dt-admin-settings', plugins_url( '/dist/css/admin.min.css', __DIR__ ), array(), $asset_data['version'] );
```

peterwilsoncc committed:
Update CSS directory.

Co-authored-by: Dharmesh Patel <dspatel44@gmail.com>

peterwilsoncc commented:
> it is `.../distributor/includes/dist/css/...` instead of `.../distributor/dist/css/...`.

Sorry, I missed that in the screen shot.

I've merged in your chagne

peterwilsoncc committed:
Fix minimum requirements in main plugin file.

iamdharmesh reviewed (dismissed):
Thanks for the changes @peterwilsoncc. All looks good to me now. :rocket: 

Thanks for your amazing work here.

dkotter commented on README.md:
```diff
@@ -169,9 +169,15 @@ Enabling this will also provide more debugging information in your error log for
-In WordPress 5.6, Application Passwords was merged into WordPress core with some limitations. From 5.6, Application Passwords is enabled by default only for live sites with HTTPS. To enable Application Passwords for development sites, you will need the following snippet:
+Application passwords are only available for live sites running over an HTTPS connection.
+
+For your local development environment will need these snippets to enable application passwords without the need for an HTTPS connection.  A local development environment is one that "can reach the internet but **is not reachable from the internet**".
```
```suggestion
For your local development environment, you will need these snippets to enable application passwords without the need for an HTTPS connection.  A local development environment is one that "can reach the internet but **is not reachable from the internet**".
```

peterwilsoncc committed:
Improve notes about local environment.

Co-authored-by: Darin Kotter <darin.kotter@gmail.com>
